compiler/rustc_passes/src/layout_test.rs

```diff
@@ -35,62 +37,64 @@ fn dump_layout_of<'tcx>(tcx: TyCtxt<'tcx>, item_def_id: LocalDefId, attr: &Attri
-                        );
+                        tcx.sess.emit_err(Size {
+                            span: tcx.def_span(item_def_id.to_def_id()),
+                            size: format!("{:?}", ty_layout.size),
+                        });
                     }
 
                     sym::homogeneous_aggregate => {
-                        tcx.sess.span_err(
-                            tcx.def_span(item_def_id.to_def_id()),
-                            &format!(
-                                "homogeneous_aggregate: {:?}",
-                                ty_layout.homogeneous_aggregate(&UnwrapLayoutCx { tcx, param_env }),
+                        tcx.sess.emit_err(HomogeneousAggregate {
+                            span: tcx.def_span(item_def_id.to_def_id()),
+                            homogeneous_aggregate: format!(
+                                "{:?}",
+                                ty_layout.homogeneous_aggregate(&UnwrapLayoutCx { tcx, param_env })
                             ),
-                        );
+                        });
                     }
 
                     sym::debug => {
-                        let normalized_ty = tcx.normalize_erasing_regions(
-                            param_env.with_reveal_all_normalized(tcx),
-                            ty,
-                        );
-                        tcx.sess.span_err(
-                            tcx.def_span(item_def_id.to_def_id()),
-                            &format!("layout_of({:?}) = {:#?}", normalized_ty, *ty_layout),
+                        let normalized_ty = format!(
+                            "{:?}",
+                            tcx.normalize_erasing_regions(
+                                param_env.with_reveal_all_normalized(tcx),
+                                ty,
+                            )
                         );
+                        let ty_layout = format!("{:#?}", *ty_layout);
+                        tcx.sess.emit_err(LayoutOf {
+                            span: tcx.def_span(item_def_id.to_def_id()),
+                            normalized_ty,
+                            ty_layout,
+                        });
                     }
 
                     name => {
-                        tcx.sess.span_err(
-                            meta_item.span(),
-                            &format!("unrecognized field name `{}`", name),
-                        );
+                        tcx.sess.emit_err(UnrecognizedField { span: meta_item.span(), name });
                     }
                 }
             }
         }
 
         Err(layout_error) => {
-            tcx.sess.span_err(
-                tcx.def_span(item_def_id.to_def_id()),
-                &format!("layout error: {:?}", layout_error),
-            );
+            tcx.sess.emit_err(Layout {
+                span: tcx.def_span(item_def_id.to_def_id()),
+                layout_error: format!("{:?}", layout_error),
```

Comment:
`LayoutError` implements `IntoDiagnostic` at the moment but the implementation of it is wrong because it isn't translatable. Do you think you could fix that? I'd also appreciate if the comment with the error added in 6e22c0a8e1799e0d14f4845060f6857d8921b283 was removed too. You should just be able to use `Spanned<LayoutError>` here instead of a new distinct type after that.

Reply:
> I'd also appreciate if the comment with the error added in [6e22c0a](https://github.com/rust-lang/rust/commit/6e22c0a8e1799e0d14f4845060f6857d8921b283) was removed too.

🤔 I don't see that comment in my branch. Perhaps it was removed in a subsequent commit?


> `LayoutError` implements `IntoDiagnostic` at the moment but the implementation of it is wrong because it isn't translatable. Do you think you could fix that?

> You should just be able to use `Spanned<LayoutError>` here instead of a new distinct type after that.

Can you take a look at what I've tried so far and let me know if I'm on the right path? It is in commit https://github.com/rust-lang/rust/pull/102110/commits/ff4397ed362ac2b6cb556b781c772d9416ec516f -- right now it is failing tests with:

```
---- [ui] src/test/ui/associated-types/issue-85103.rs stdout ----
diff of stderr:

-       error: layout error: NormalizationFailure(<[E] as std::borrow::ToOwned>::Owned, Type(<[E] as std::borrow::ToOwned>::Owned))
+       error: unable to determine layout for `<[E] as ToOwned>::Owned` because `<[E] as ToOwned>::Owned` cannot be normalized
2         --> $DIR/issue-85103.rs:6:1
3          |
4       LL | type Edges<'a, E> = Cow<'a, [E]>;
```